File: wgsn/src/test/java/wgsn/examples/java/automation/core/WebDriverExtensions.java
package wgsn.examples.java.automation.core;

import org.openqa.selenium.Alert;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.remote.RemoteWebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;
import wgsn.examples.java.automation.tests.SharedDriver;

import java.util.Collection;

/**
 * Created with IntelliJ IDEA.
 * User: Yomi Folami
 * Date: 05/03/13
 * Time: 18:36
 * To change this template use File | Settings | File Templates.
 */
public class WebDriverExtensions extends SharedDriver {
    private static RemoteWebDriver WEB_DRIVER = SharedDriver.WEBDRIVER;
    private static Actions action = new Actions(current());
    private static WebDriverWait wait = new WebDriverWait(current(), 10);

    public static WebDriver current() {
        return WEB_DRIVER;
    }

    /*
        Custom page element finders.
     */
    public static WebElement findElementByCssSelector(String css) {
        return current().findElement(By.cssSelector(css));
    }

    public static Collection<WebElement> findElementsByCssSelector(String css) {
        return current().findElements(By.cssSelector(css));
    }

    public static WebElement findElementByLinkText(String link) {
        return current().findElement(By.linkText(link));
    }

    /*
        Custom actions.
     */
    public static void clickOn(WebElement element) {
        element.click();
    }

    public static void typeText(WebElement element, String _string) {
         if(!element.getAttribute("value").isEmpty()) {
             element.clear();
         }
         element.sendKeys(_string);
    }

    public static void navigateTo(WebDriver driver, String url) {
        driver.navigate().to(url);
    }

    public static void mouseOver(WebElement element) {
        action.moveToElement(element).perform();
    }

    public static void select(WebElement element, String _string) {
        Select select = new Select(element);
        select.selectByVisibleText(_string);
    }

    public static void dragAndDrop(WebElement source, WebElement target) {
         action.dragAndDrop(source, target).perform();
    }

    public static void rightClick(WebElement element) {
        action.contextClick(element).perform();
    }

    /*
        waits for page to load and verifies page title.
     */
    public static void verifyPageTitle(String title) {
        if(!wait.until(ExpectedConditions.titleContains(title))) {
            System.err.println("Page title doesn't contain {" + title + "} but {"
                    + WEBDRIVER.getTitle() + "}");
        }
    }

    /*
        waits until element is visible on screen.
     */
    public static void waitForElement(WebElement element) {
        if(wait.until(ExpectedConditions.visibilityOf(element)) != null) return;
    }

    public static boolean isElementDisplayed(WebElement element) {
        return element.isDisplayed();
    }

    // switch driver to frame
    public static void switchToFrame(int frameId) {
        current().switchTo().frame(frameId);
    }

    /*
        alert and pop-up methods
     */
    public static Alert switchToAlert() {
        Alert alert = current().switchTo().alert();
        return alert;
    }

    public static void chooseOkOnNextConfirmation() {
        Alert alert = switchToAlert();
        alert.accept();
    }

    public static void chooseCancelOnNextConfirmation() {
        Alert alert = switchToAlert();
        alert.dismiss();
    }
}

File: dotnet/Scribestar.ET.Regression/Scribestar.ET.Regression/Core/Support/WebDriverExtensions.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Interactions;
using OpenQA.Selenium.Remote;
using OpenQA.Selenium.Support.UI;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace Scribestar.ET.Regression.Core.Support
{
    public static class WebDriverExtensions
    {
        private static readonly RemoteWebDriver WebDriver = SharedDriver.DriverInstance.WebDriver;
        private static readonly WebDriverWait Wait = new WebDriverWait(CurrentPage(), SharedDriver.TimeToWait);
        private static IAlert Alert;

        public static RemoteWebDriver CurrentPage()
        {
            return WebDriver;
        }

        /// <summary>
        ///     Custom page element finders.
        /// </summary>
        public static IWebElement FindWebElementByLinkText(string linkText)
        {
            return WebDriver.FindElementByLinkText(linkText);
        }

        public static IWebElement FindWebElementByCssSelector(string css)
        {
            return WebDriver.FindElementByCssSelector(css);
        }

        public static ReadOnlyCollection<IWebElement> FindWebElementsByCssSelector(string css)
        {
            return WebDriver.FindElementsByCssSelector(css);
        }

        public static ReadOnlyCollection<IWebElement> FindWebElementsByTagName(string tagName)
        {
            return WebDriver.FindElementsByTagName(tagName);
        }

        /// <summary>
        ///     Custom interaction handlers
        /// </summary>
        public static void NavigateTo(this IWebDriver driver, string url)
        {
            driver.Navigate().GoToUrl(url);
        }

        public static void TypeText(this IWebElement element, string _string)
        {
            if (!element.GetAttribute("value").Length.Equals(0))
                element.Clear();

            element.SendKeys(_string);
        }

        public static void ClickOn(this IWebElement element)
        {
            if (element != null) element.Click();
        }

        public static void Select(this IWebElement element, string _string)
        {
            var select = new SelectElement(element);
            @select.SelectByText(_string);
        }

        public static void MouseOver(this IWebElement element)
        {
            var action = new Actions(CurrentPage());
            @action.MoveToElement(element).Perform();
        }

        public static bool IsElementDisplayed(this IWebElement element)
        {
            return element.Displayed;
        }

        public static void ClickButton(string css)
        {
            var element = FindWebElementByCssSelector(css);
            element.ClickOn();
        }


        /// <summary>
        ///     Custom wait handler
        /// </summary>
        public static void WaitForElement(string _string)
        {
            if (Wait.Until(ExpectedConditions.ElementIsVisible(By.CssSelector(_string))) != null) return;

            var reason = "Element: {" + _string + "} is not visible or loaded on " + CurrentPage().Url;
            var ex = new NoSuchElementException();
            //ex.Should().BeAssignableTo<ExpectedConditions>(@reason);
        }

        public static void WaitForAlert()
        {
            var alert = (IWebElement)Alert;
            if (alert.IsElementDisplayed()) return;

            var reason = "Alert popup failed to display on " + CurrentPage().Url;
        }


        /// <summary>
        ///     Custom alert and pop up handler
        /// </summary>
        public static IAlert SwitchToAlert()
        {
            var alert = CurrentPage().SwitchTo().Alert();
            return @alert;
        }

        public static void ChooseOkOnNextConfirmation()
        {
            var accept = SwitchToAlert();
            @accept.Accept();
        }

        public static void ChooseCancelOnNextConfirmation()
        {
            var cancel = SwitchToAlert();
            @cancel.Dismiss();
        }


    }
}
